### ISSUE TITLE
FR: Limit Font List Based on Book Language/Typography Rules

### ISSUE THREAD

**Issue by Bluemoondragon07** [koreader/koreader#11901]
So, this is a minor idea that I think would be cool. How many of us read in multiple languages? I am one of them. I have a lot of English fonts and some Japanese Fonts. It is hard to scroll through all my English fonts to find a Japanese one. 

Now, that is mostly my problem for being so darn addicted to collecting English fonts. But, imagine how convenient it would be if I could choose to filter these fonts by the language of the book? 

Lang=jpn? Automatically show me only Japanese fonts. 
And then, in English books, I also won't see my Japanese fonts. 

But now, what about bilingual books? Surely we'd need fallback fonts in a different language for some of these? Perhaps we can implement another option—instead of filtering the fonts 100%, we can just sort them differently. In a Japanese book, show the Japanese fonts first, then show fonts for other languages. In an English book, show English fonts first.

And how would we know which fonts are for what language? Read the `dlng` property of each ttf file. Perhaps even `slng` can optionally be included in the filter. If a font does not have these properties, maybe we can even check for unicode ranges, but I do not know if this is possible with the app.

Anyways, please consider!

**Comment by Frenzie** [comment]
If implemented I'd definitely go with the slightly different sorting order over hiding anything, as a type of precaution.

**Comment by poire-z** [comment]
Too complex, hard to explain to users, too niche with a population of one user :)

If I understand right, you would just want to filter (or sort) the fonts shown in the menu, and nothing else ? This could probably be doable with a user patch, patching bits of apps/modules/readerfont.lua. (calling ReaderFont:setupFaceMenuTable() would probably need to be moved from :init() to an added :onReaderReady() so after we know the book language).

About detecting which font support what, it's always risky. I'm sure some of your dear japanese fonts are old and don't have all the OpenType tables you expect.
So, simpler would be to arrange the fonts in multiple subdirectories, `ja/` `en/` `others/` and have the user patch filter them from the book language and the font path.

**Comment by Bluemoondragon07** [comment]
True, sounds complex. Although, I do not think it would be hard to explain. It could be one toggle: "Sort by Language" in fonts. As for niche, we certainly aren't a majority, but I think tons of people read in more than one language. I guess that still counts as niche, but it's a decent percentage. 

I guess I'll try the patch thing, but I'll have to learn Lua first 😬. Was planning on doing that at some point anyway.

**Comment by Frenzie** [comment]
> So, simpler would be to arrange the fonts in multiple subdirectories, `ja/` `en/` `others/` and have the user patch filter them from the book language and the font path.

Sorting by directory structure sounds more dependable indeed.

**Comment by mergen3107** [comment]
Profiles + Gestures is your answer

**Comment by mergen3107** [comment]
Just set up all your fonts as a profile each, then put them all in Quick Menu and summon them via gesture.

**Comment by offset-torque** [comment]
Main problem is that the number of fonts exposed in the list is too low, which means lots of horizontal swiping. If our UI was more flexible, we could easily fit 3 times more fonts in that empty space. But probably this is impossible:

![font_box1](https://github.com/koreader/koreader/assets/84420582/1bd537bd-0960-4ebb-bb72-e61c63d6ebd4)

**Comment by mergen3107** [comment]
The number of items per page was made so the body text is visible when changing fonts :D

**Comment by offset-torque** [comment]
Yeah I am aware, that's why I didn't change the height in the mockup, just the width ;)

**Comment by Commodore64user** [comment]
If only there was a font menu... ;)

**Comment by jonnyl2** [comment]
Instead of additional fonts per menu page, I'd rather see the same font in different available styles (e.g., italics and bold). Some fonts are rather ugly in italics, and others don't even have an italics version (and fake italics looks the _worst_). And there's no way of telling from the menu :-/

**Comment by Frenzie** [comment]
Try font settings → generate font test document.

**Comment by jonnyl2** [comment]
Yep, I'm aware of that, thanks. But it's not quite the same thing as a live preview.

**Comment by Bluemoondragon07** [comment]
Yes, I guess what we really want is a sort by folders option for fonts. Now that I think about it, most of my Japanese fonts are in a `japanese/` subfolder anyway.

 This would also allow for more customization. People may already categorize fonts in serif, mono, and sans subfolders; others may categorize their fonts by genre (fantasy, scholarly, sci-fi). I currently do this with profiles but I have soooo many fonts—doing a profile for each font one by one takes hours of my life—but that's also, again, my problem for being so darn obssessed with adding fonts to my collection and then categorizing them...

And if this feature is of interest, maybe it can be taken a step further and include filters for each subfolder. Maybe. Sorting would be very sufficient, so filtering would just be an unnecessary but cool extra.
